pallets/polkadex-ido/src/lib.rs

```diff
@@ -506,122 +225,99 @@ pub mod pallet {
             <InfoProjectTeam<T>>::insert(team, round_id);
             Self::deposit_event(Event::FundingRoundRegistered(round_id));
             Ok(())
         }
 
-        /// Project team whitelists investor for the given round for the given amount.
+        /// Invest in a funding round
         ///
         /// # Parameters
         ///
         /// * `round_id`: Funding round id
-        /// * `investor_address`: Investor
-        /// * `amount`: The max amount that investor will be investing in tokenB
+        /// * `amount`: BalanceOf<T>
         #[pallet::weight((10_000, DispatchClass::Normal))]
-        pub fn whitelist_investor(origin: OriginFor<T>, round_id: T::Hash, investor_address: T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
-            let team: T::AccountId = ensure_signed(origin)?;
-            ensure!(!<InfoFundingRound<T>>::contains_key(&round_id), Error::<T>::FundingRoundNotApproved);
-            ensure!(<WhitelistInfoFundingRound<T>>::contains_key(&round_id), Error::<T>::FundingRoundDoesNotExist);
-            let funding_round = <WhitelistInfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundNotApproved)?;
-            ensure!(team.eq(&funding_round.creator), <Error<T>>::NotACreater);
-            ensure!(<InfoInvestor<T>>::contains_key(&investor_address), <Error<T>>::InvestorDoesNotExist);
+        pub fn invest(origin: OriginFor<T>, round_id: T::Hash, amount: BalanceOf<T>) -> DispatchResult {
+            let investor_address: T::AccountId = ensure_signed(origin)?;
+            ensure!(<InfoFundingRound<T>>::contains_key(&round_id), Error::<T>::FundingRoundDoesNotExist);
+            let mut funding_round = <InfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundNotApproved)?;
+            ensure!(Self::can_withdraw(funding_round.token_b,&investor_address, amount.saturated_into()).is_ok(), Error::<T>::BalanceInsufficientForInteresetedAmount);
+            let amount_in_token_a = if T::OnePDEX::get().saturated_into::<BalanceOf<T>>() >= funding_round.token_a_price_per_token_b {
+                funding_round.token_a_price_per_1e12_token_b().saturating_reciprocal_mul(amount) 
+            } else {
+                amount / funding_round.token_a_price_per_1e12_token_b_balance()
+            };
+            ensure!(amount_in_token_a <= funding_round.max_allocation && amount_in_token_a >= funding_round.min_allocation, Error::<T>::NotAValidAmount);
             let current_block_no = <frame_system::Pallet<T>>::block_number();
-            ensure!(current_block_no < funding_round.close_round_block && current_block_no >= funding_round.start_block, <Error<T>>::NotAllowed);
-            <WhiteListInvestors<T>>::insert(round_id, investor_address.clone(), amount);
-            Self::deposit_event(Event::InvestorWhitelisted(round_id, investor_address));
+            ensure!(current_block_no >= funding_round.start_block && current_block_no < funding_round.close_round_block, <Error<T>>::NotAllowed);
```

Comment:
Move this check to before amount calculation

Reply:
Updated ✅